codec.go

```diff
@@ -20,11 +20,46 @@ package grpc
 
 import (
 	"math"
+	"strings"
 	"sync"
 
 	"github.com/golang/protobuf/proto"
 )
 
+var (
+	registeredProtoCodec Codec = protoCodec{}
+	registeredCodecs           = map[string]Codec{"proto": registeredProtoCodec}
+)
+
+// RegisterCodec registers the provided Codec for use with all gRPC clients
+// and servers.
+//
+// The Codec will be stored and looked up by result of it's String() method,
+// which should match the content-subtype of the encoding handled by the Codec.
+// This is case-insensitive, and is stored and looked up as lowercase.
+// If the result of calling String() is an empty string, RegisterCodec will
+// panic. See Content-Type on https://grpc.io/docs/guides/wire.html#requests
+// for more details.
+//
+// By default, a Codec for "proto" is registered.
+//
+// RegisterCodec should only be called from init() functions, and is not
+// thread-safe. Codecs can be overwritten in the registry, including the
+// default "proto" Codec.
+func RegisterCodec(codec Codec) {
```

Comment:
I need more direction here - will `google.golang.org/grpc` depend on `google.golang.org/grpc/encoding`? Should I add another public method `GetRegisteredCodec(contentSubtype string) Codec` to `encoding`? Also maybe `GetRegisteredProtoCodec() Codec` which must return nil?

If `Codec` is still in `google.golang.org/grpc`, this will result in a cyclical dependency, as I need access to the registered codecs there, but `encoding` will rely on the base package.

Reply:
> I need more direction here - will `google.golang.org/grpc` depend on `google.golang.org/grpc/encoding`?

Yes

> Should I add another public method `GetRegisteredCodec(contentSubtype string) Codec` to `encoding`? Also maybe `GetRegisteredProtoCodec() Codec` which must return nil?

Yes.  (I prefer `GetCodec(string) Codec`, though, to be less verbose.)

> If `Codec` is still in `google.golang.org/grpc`, this will result in a cyclical dependency, as I need access to the registered codecs there, but `encoding` will rely on the base package.

The cyclical dependency is avoided by copying the `Codec` definition to `encoding` and referencing that one from both the `grpc` and `encoding` packages.  The old one only exists temporarily in case external users are referencing it.  Maybe:

```go
package grpc

var _ Codec = encoding.Codec(nil)
```

To paranoidly guarantee they are identical?

Reply:
This is done, however i did not `s/Codec/encoding.Codec/` in `grpc`. I think this should be left to another PR as if this causes problems, it should be independently roll-backable.

Note it's your repository so this is up to you :-) but I am very against not following SemVer and breaking the API between major releases, which moving `Codec` from `grpc` to a new `encoding` package is. This has caused us and others problems in the past.

Reply:
>  I think this should be left to another PR as if this causes problems, it should be independently roll-backable.

That's fine, I agree with that premise.

But, I'm not sure I understand where you're seeing the risk of breakage.  I think the only possible breakage is if someone is doing something like this:

```go
var codecDialOption = grpc.WithCodec
```

and then in a test:

```go
codecDialOption = func(grpc.Codec) grpc.DialOption {
  // ???
}
```

Otherwise, these are identical interfaces, so anything that satisfies one satisfies the other.  Let me know if you see something I'm overlooking.

> Note it's your repository so this is up to you :-) but I am very against not following SemVer

I'm pretty sure my hands are tied on our release process and versioning, but I can put out some feelers.